### ISSUE TITLE
Using HiltAndroidApp annotation within the library module

### ISSUE THREAD

**Issue by grkep** [google/dagger#3651]
Hi,
I'm trying to integrate Hilt in a multi-module android project(we don't use feature modules).
Here is a current structure of our project
- App Module
- Lib Module 1 (This library has Application class `MyApplication1` for app 1)
- Lib Module 2 (This library has Application class `MyApplication2` for app 2)
- Lib module 3 (common stuff)
- Up to Lib module 100

If I put `@HiltAndroidApp` annotation on `MyApplication1` I get this build error:
```
Application class annotated with @HiltAndroidApp has to be defined in an android application project
  [Hilt] Processing did not complete. See error above for details.
```

To address this issue, if I move the application classes `MyApplication1` and `MyApplication2` to the App module, I'll start getting circular dependency errors(because a library project cannot depend on App project). I've dozens of libraries and they have code(e.g. `lateinit var app: MyApplication1`, access app specific properties, etc.) which needs a reference to the Application object, hence I cannot have independent Application classes. If you are suggesting to refactor our code base to have independent Application module, then that would likely be a year long project(given how large our code base is)

I also tried creating a separate app module(not library module) holding a base Application and made my `MyApplication1` and `MyApplication2` extend it, but I can't do that successfully, because Android Studio doesn't like a `library-module` depend on an `application-module`. And having a library-module depend on app-module doesn't sound like a good pattern, and going to cause other issues anyway.

I came across this [issue](https://github.com/google/dagger/issues/3400) but the suggested solution on the issue won't work in my case(due to reasons mentioned above).

Will Hilt ever support annotating an Application class with @HiltAndroidApp within the library modules?

**Comment by hubtwork** [comment]
I think it's weird that architecture your project is following. As you said, Application module is not a app end point.
you said,
App Module , Lib1 ( App endpoint 1) , Lib1 ( App endpoint 2) , and so on.
I think that your app module is not a endpoint.
I'v been multi-module & multi-application project for some works, we have modulizing architecture like below.
```
- Android App1 ( for customers )
- Android App2 ( for companies )
- Android App3 ( for admin )
- Android Lib ( core - App ) 
- and so on for each roles in clean architectures ( 10 + )
```
**Android Lib ( core - App )** has base application class for DI and initialize each manager/ helpers
```
abstract class CoreApp: Application() {
    companion object {
        lateinit var instance: CoreApp
    }
   ...
}
```
and each application modules for enpoint app has just application implementation of `CoreApp` like below.
```
(in App1 module) 
@HiltAndroidApp
class CustomerApp: CoreApp() { ... }
(in App2 module)
@HiltAndroidApp
class CompanyApp: CoreApp() { ... }
(in App3 module)
@HiltAndroidApp
class AdminApp: CoreApp() { ... }
```

It's proper `@HiltAndroidApp` that hilt points is proper for android architecture.
**Application** class is not means entrypoint with hilt, **@HiltAndroidApp** annotated class which extends Application is entrypoint in hilt.

So if you want build project with hilt and multi module / app project, you should abstract base application class to Lib Module, and implements it from each `Android App modules`.
It seems that `circular dependency errors` you said is not a problem, your dependency direction ( app -> lib1, lib2 ) is problem.
Just refactor App module to Lib module and refactor each Lib modules to App modules which implements the `BaseApplication` in Lib module ( which is in App module for your current project ).

**Comment by grkep** [comment]
@hubtwork thanks for taking time to respond.
I get that refactoring is a correct next step. I tried some quick refactoring, and the errors disappeared, and I was able to get some basic injection working. One gotcha, I would like to share w.r.t. multi-module setup is the hilt dependency. Such that, I had to add hilt dependencies `implementation "com.google.dagger:hilt-android:$hilt_version"    kapt "com.google.dagger:hilt-android-compiler:$hilt_version"` to the `build.gradle` of each module(where I wanted injection), in addition to adding the dependencies in the root build.gradle. A common library with dependencies in the form of `api "com.google.dagger:hilt-android:$hilt_version"  kapt "com.google.dagger:hilt-android-compiler:$hilt_version"` didn't work(Got null references for dependency in this case, no compile time errors). If someone has any idea about not repeating the dependencies, please do share!

**Comment by hubtwork** [comment]
I think that if your libraries are which is just implementation all abstractions with POJO or don't use any hilt annotations, it don't need any hilt dependencies.
Because in your project ( i'm not sure whole structure ), it seems to have aggregator or di-role module. ( in my case, we have whole DI for each libs in each App modules )
So except for UI modules ( which needs `@HiltViewModel` for Dependency Injection to Viewmodel ) and aggregator, you don't need to dependent on hilt.

p.s. 
Reason why i wrote upon, i depends hilt only ( App modules , UI feature modules , hilt extension module ). Other 20+ modules are non-hilt modules.

**Comment by grkep** [comment]
@hubtwork yeah, that's correct. If module doesn't use hilt I don't need to add dependencies. I just wanted to know if there is any way to avoid repeating dependencies each time, or it's a bug in hilt(since I'm already using other dependencies through a common lib module). This would help to keep the code clean and DRY.

**Comment by hubtwork** [comment]
@grkep hmm.. in my knowledge, it isn't bug. Hilt is android-wrapper for dagger, and dagger designed with some inconvenience ( as you said, repeating dependencies,, etc ) but with best convenience for Dependency Injection by source-auto generating processors.
